how-it-works/1_protocol/04_execution.subpage.md

```diff
@@ -5,8 +5,92 @@ shareImage: /img/how-it-works/network-nervous-system-nns.600.jpg
+However, all the nodes of the subnet can concurrently execute different queries because queries are not executed in a replicated way.
+Query throughput of a subnet thus increases linearly with an increasing number of nodes in the subnet, while the update call performance decreases with an increasing number of nodes.
+
+Queries are similiar to read operations on a local or cloud Ethereum node on the Ethereum blockchain.
+A dApp should use queries for non-critical operations only.
+Whenever an information item to be read is critical, e.g., financial data based on which decisions are made, update calls should be used to obtain such information as the response of an update call is certified by the subnet with a BLS threshold signature and verifiable with the subnet's public key.
+
+## Memory Handling
+
+Management of the canister bytecode and state (collectively memory) is one of the key responsibilities of the execution layer.
+The replicated state that can be held by a single subnet is not bounded by the available RAM in the node machines, but rather by the available SSD storage.
+Available RAM, however, impacts the performance of the subnet, particularly the access latency of memory pages.
+This depends a lot on the access patterns of the workload, however, – much like in traditional computer systems.
+
+Our node machines are equipped with tens of terabytes of high-end SSD storage and over half a terabyte of RAM to be able to hold large amounts of replicated canister state and Wasm code and achieve good performance when accessing memory.
+While the execution layer maintains the memory data structures and updates them accordingly to reflect modifications to memory pages, the state management component of message routing is responsible for certifying (parts of) the state through the subnet at the end of a round or once every (much longer) checkpointing interval.
+
+Memory pages representing canister state are persisted to SSD by the execution layer, without canister programmers needing to take care of this.
+Having all memory pages transparently persisted enables *orthogonal persistence* and frees the smart contract programmers from reading from and writing to storage as on other blockchains or as in traditional IT systems.
+This dramatically simplifies smart contract implementation and helps reduce the TCO of a dApp and go to market faster.
+Programmers can always have the full canister smart contract state on the heap or in stable memory.
+The difference is that the heap is cleared on updates of the canister code, while stable memory remains stable throughout updates, hence its name.
+Any state on the heap that is to be preserved through a canister update must be transferred to stable memory by a canister programmer before an update and restored from there after the update.
+Best practices are that large canister state be held directly in stable memory to avoid shuffling around large amounts of storage before and after each upgrade.
```

Comment:
I don't know if this is always a best practice -- perhaps we should not mention this at this point. I think best practices could be grouped once we have a nice collection of such practices.

Reply:
Sitting with Dieter - thanks for the suggestion, we'll note for a best practices doc!